ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/block/BlockManager.java

```diff
@@ -124,87 +133,119 @@ private void importBlockIgnoringResult(final SignedBeaconBlock block) {
-                    // the pending pool before this check happens to avoid race conditions between
-                    // performing the check and the parent importing.
-                    if (recentChainData.containsBlock(block.getParentRoot())) {
-                      pendingBlocks.remove(block);
-                      importBlockIgnoringResult(block);
-                    }
-                    break;
-                  case BLOCK_IS_FROM_FUTURE:
-                    futureBlocks.add(block);
-                    break;
-                  case FAILED_EXECUTION_PAYLOAD_EXECUTION_SYNCING:
-                    LOG.warn("Unable to import block: Execution Client is still syncing");
-                    break;
-                  case FAILED_EXECUTION_PAYLOAD_EXECUTION:
-                    LOG.error(
-                        "Unable to import block: Execution Client communication error.",
-                        result.getFailureCause().orElse(null));
-                    break;
-                  default:
-                    LOG.trace(
-                        "Unable to import block for reason {}: {}",
-                        result.getFailureReason(),
-                        block);
-                    dropInvalidBlock(block);
-                }
-              }
-            });
+  private boolean blockIsInvalid(final SignedBeaconBlock block) {
+    return invalidBlockRoots.containsKey(block.getRoot())
+        || invalidBlockRoots.containsKey(block.getParentRoot());
   }
 
-  private boolean shouldImportBlock(final SignedBeaconBlock block) {
-    if (blockIsKnown(block)) {
-      return false;
-    }
-    if (blockIsInvalid(block)) {
-      dropInvalidBlock(block);
-      return false;
-    }
-    return true;
+  private Optional<BlockImportResult> handleInvalidBlock(final SignedBeaconBlock block) {
+    Optional<BlockImportResult> blockImportResult =
+        Optional.ofNullable(invalidBlockRoots.get(block.getRoot()))
+            .or(
+                () -> {
+                  if (invalidBlockRoots.containsKey(block.getParentRoot())) {
+                    return Optional.of(FailedBlockImportResult.FAILED_DESCENDANT_OF_INVALID_BLOCK);
+                  }
+                  return Optional.empty();
+                });
+
+    blockImportResult.ifPresent(result -> dropInvalidBlock(block, result));
+
+    return blockImportResult;
   }
 
-  private boolean blockIsKnown(final SignedBeaconBlock block) {
-    return pendingBlocks.contains(block)
-        || futureBlocks.contains(block)
-        || recentChainData.containsBlock(block.getRoot());
+  private Optional<BlockImportResult> handleChildOfFailedExecutionPayloadBlock(
+      final SignedBeaconBlock block) {
+    if (executionFailedBlockRoots.contains(block.getParentRoot())) {
+      executionFailedBlockRoots.add(block.getRoot());
+      pendingBlocks.add(block);
+      return Optional.of(BlockImportResult.FAILED_DESCENDANT_OF_FAILED_EXECUTION_PAYLOAD_BLOCK);
```

Comment:
yes true. It just adds a specific reason for this particular case, instead of returning UNKNOWN_PARENT.
Let's reduce unnecessary lines of code, if we actually do nothing special with that return code.